modules/twinkleblock.js

```diff
@@ -0,0 +1,870 @@
+      { label: 'Promotional username, hard block', value: 'uw-spamublock' },
+      { label: 'Promotional username, soft block', value: 'uw-softerblock' },
+      { label: 'Similar username soft block', value: 'uw-ublock-double' },
+      { label: 'Username violation, soft block', value: 'uw-ublock' },
+      { label: 'Username violation, hard block', value: 'uw-uhblock' },
+      { label: 'Username impersonation hard block', value: 'uw-uhblock-double' },
+      { label: 'Username represents a famous person, soft block', value: 'uw-ublock-famous' },
+      { label: 'Username represents a non-profit, soft block', value: 'uw-causeblock' },
+      { label: 'Username violation, vandalism-only account', value: 'uw-vaublock' }
+    ]
+  },
+  {
+    label: 'Templated reasons',
+    list: [
+      { label: 'anonblock', value: 'anonblock' },
+      { label: 'anonblock - likely a school', value: 'anonblock - school' },
+      { label: 'blocked proxy', value: 'blocked proxy' },
+      { label: 'CheckUser block', value: 'CheckUser block' },
+      { label: 'checkuserblock-account', value: 'checkuserblock-account' },
+      { label: 'checkuserblock-wide', value: 'checkuserblock-wide' },
+      { label: 'colocationwebhost', value: 'colocationwebhost' },
+      { label: 'oversightblock', value: 'oversightblock' },
+      { label: 'school block', value: 'school block' },
+      // { label: 'rangeblock', value: 'rangeblock' },
+      { label: 'tor', value: 'tor' },
+      { label: 'webhostblock', value: 'webhostblock' }
+    ]
+  }
+];
+
+Twinkle.block.callback.filtered_block_groups = function twinkleblockCallbackFilteredBlockGroups(show_template) {
+  return $.map(Twinkle.block.blockGroups, function(value) {
+    var list = $.map(value.list, function(v) {
+        var blockSettings = Twinkle.block.blockPresetsInfo[v.value];
+        var registrationRestrict = blockSettings.forRegisteredOnly ? Twinkle.block.isRegistered : (blockSettings.forAnonOnly ? !Twinkle.block.isRegistered : true);
+        if (!blockSettings.excludeFromTemplates && registrationRestrict) {
+          return {
+            label: (show_template ? '{{' + v.value + '}}: ' : '') + v.label,
+            value: v.value
+          };
+        }
+      });
+    if (list.length) return {
+        label: value.label,
+        list: list
+      };
+  });
+};
+
+Twinkle.block.callback.change_preset = function twinkleblockCallbackChangePreset(e) {
+  var $form = $(e.target.form), key = $form.find('[name=category]').val();
+  if (!key) return;
+
+  $form.find('[name=sub_group]').val(key);
+  var preset = Twinkle.block.blockPresetsInfo[key];
+  Twinkle.block.callback.update_form(e, preset);
+  Twinkle.block.callback.change_subcategory(e);
+};
+
+Twinkle.block.callback.change_expiry = function twinkleblockCallbackChangeExpiry(e) {
+  var $expiry = $(e.target.form).find('[name=expiry]');
+  if ($(e.target).val() === 'custom') {
+    $expiry.parent().show();
```

Comment:
Consider using the utility method `Morebits.quickForm.setElementVisibility` instead.